### PR TITLE
Handling broken layout due to hidden null value appears beyond the view group

### LINKED ISSUE #4601: Apply conditional encoding to make sure null values are not off the view group
Null values are no longer filtered, and thus can be off-chart when the axis does not contain zero.  

The trellis histogram is one example that demonstrates this problem. 

image

### PR DESCRIPTION
**Pull request by kanitw**
Merge #4604 first.  

Fix #4601

### CODE REVIEW

**Comment by kanitw** [comment]
@domoritz I actually just realized that another solution is to turn "clamp" on by default.

**Comment by kanitw** [comment]
This is a bit more complicated for the case needed, but will affect fewer examples.  Changing clamp will affect all examples.    Any preference?

**Comment by domoritz** [comment]
If you zoom, clamp won't fix this issue, right?

**Comment by kanitw** [comment]
> If you zoom, clamp won't fix this issue, right?

It will, but basically, you cannot turn on clamp without facing this problem anymore.

So I think it's better to do this predicate things so people still have freedom to use clamp or not.  

Plus, some code here would be a good infrastructure for even better null handling https://github.com/vega/vega-lite/issues/2544.

**Comment by kanitw** [comment]
image

File: src/compile/scale/component.ts [hunk 1]
```diff
@@ -24,6 +25,20 @@ export class ScaleComponent extends Split<ScaleComponentProps> {
     );
     this.setWithExplicit('type', typeWithExplicit);
   }
+
+  /**
+   * Whether the scale definitely includes zero in the domain
+   */
+  public get domainDefinitelyIncludeZero() {
```
Comment (domoritz) [inline src/compile/scale/component.ts]:
```suggestion
  public get domainDefinitelyIncludesZero() {
```

File: src/vega.schema.ts [hunk 2]
```diff
@@ -68,6 +68,7 @@ export interface VgValueRef {
   mult?: number;
   offset?: number | VgValueRef;
   band?: boolean | number | VgValueRef;
+  test?: string;
```
Comment (domoritz) [inline src/vega.schema.ts]: Can we use the type from Vega instead to avoid problems like this in the future?

File: src/compile/mark/valueref.ts [hunk 3]
```diff
@@ -88,13 +145,22 @@ export function position2({
...
+    defaultRef
+  });
 }
 
 export function getOffset(channel: 'x' | 'y' | 'x2' | 'y2', markDef: MarkDef) {
-  const offsetChannel = channel + 'Offset';
-  // TODO: in the future read from encoding channel too
+  const offsetChannel = (channel + 'Offset') as 'xOffset' | 'yOffset' | 'x2Offset' | 'y2Offset'; // Need to cast as the type can't be inferred automatically
 
+  // TODO: in the future read from encoding channel too
```
Comment (domoritz) [inline src/compile/mark/valueref.ts]: Why not now?

Comment (kanitw) [reply src/compile/mark/valueref.ts]: There is no offset channel

File: src/vega.schema.ts (hunk 2 above)
Comment (kanitw) [reply src/vega.schema.ts]: Yes, but it’s beyond the scope of this PR.  Please send a follow up fix.

**Comment by domoritz** [comment]
Cool, lgtm

**Pull request merged**